core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -100,48 +92,136 @@ private[spark] class MemoryStore(
+    var unrollMemoryUsedByThisBlock = 0L
+    // Underlying vector for unrolling the block
+    var vector = new SizeTrackingVector[Any]
+
+    // Request enough memory to begin unrolling
+    keepUnrolling = reserveUnrollMemoryForThisTask(blockId, initialMemoryThreshold)
+
+    if (!keepUnrolling) {
+      logWarning(s"Failed to reserve initial memory threshold of " +
+        s"${Utils.bytesToString(initialMemoryThreshold)} for computing block $blockId in memory.")
+    } else {
+      unrollMemoryUsedByThisBlock += initialMemoryThreshold
+    }
+
+    // Unroll this block safely, checking whether we have exceeded our threshold periodically
+    while (values.hasNext && keepUnrolling) {
+      vector += values.next()
+      if (elementsUnrolled % memoryCheckPeriod == 0) {
+        // If our vector's size has exceeded the threshold, request more memory
+        val currentSize = vector.estimateSize()
+        if (currentSize >= memoryThreshold) {
+          val amountToRequest = (currentSize * memoryGrowthFactor - memoryThreshold).toLong
+          keepUnrolling = reserveUnrollMemoryForThisTask(blockId, amountToRequest)
+          if (keepUnrolling) {
+            unrollMemoryUsedByThisBlock += amountToRequest
           }
+          // New threshold is currentSize * memoryGrowthFactor
+          memoryThreshold += amountToRequest
+        }
+      }
+      elementsUnrolled += 1
+    }
+
+    if (keepUnrolling) {
+      // We successfully unrolled the entirety of this block
+      val arrayValues = vector.toArray
+      vector = null
+      val entry = if (level.deserialized) {
+        val sizeEstimate = SizeEstimator.estimate(arrayValues.asInstanceOf[AnyRef])
+        new MemoryEntry(arrayValues, sizeEstimate, deserialized = true)
+      } else {
+        val bytes = blockManager.dataSerialize(blockId, arrayValues.iterator)
+        new MemoryEntry(bytes, bytes.limit, deserialized = false)
+      }
+      val size = entry.size
+      val acquiredStorageMemory = {
+        if (unrollMemoryUsedByThisBlock == size) {
+          true
+        } else if (unrollMemoryUsedByThisBlock < size) {
+          memoryManager.acquireStorageMemory(blockId, size - unrollMemoryUsedByThisBlock)
         } else {
-          val bytes = blockManager.dataSerialize(blockId, arrayValues.iterator)
-          if (tryToPut(blockId, () => bytes, bytes.limit, deserialized = false)) {
-            Right(bytes.limit())
-          } else {
-            Left(arrayValues.toIterator)
-          }
+          memoryManager.releaseUnrollMemory(unrollMemoryUsedByThisBlock - size)
+          true
         }
-      case Right(iteratorValues) =>
-        Left(iteratorValues)
+      }
+      if (acquiredStorageMemory) {
+        // We acquired enough memory for the block, so go ahead and put it
+        memoryManager.synchronized {
+          unrollMemoryMap(currentTaskAttemptId()) -= unrollMemoryUsedByThisBlock
```

Comment:
it might be better to add an assert here:
```
val taskAttemptId = currentTaskAttemptId()
assert(unrollMemoryMap.contains(taskAttemptId))
assert(unrollMemoryMap(taskAttemptId) >= unrollMemoryUsedByThisBlock)
unrollMemoryMap(taskAttemptId) -= unrollMemoryUsedByThisBlock
```
not super necessary but it seems safer in case we screwed up the accounting elsewhere